src/main/java/org/candlepin/auth/CloudRegistrationAuth.java

```diff
@@ -271,19 +278,114 @@ public String generateRegistrationToken(Principal principal, CloudRegistrationIn
+     *     The registration information to validate
+     *
+     * @throws UnsupportedOperationException
+     *     if the current Candlepin configuration does not support cloud registration
+     *
+     * @throws CloudRegistrationAuthorizationException
+     *     if cloud registration is not permitted for the cloud provider or account holder specified by
+     *     the cloud registration details
+     *
+     * @throws MalformedCloudRegistrationException
+     *     if the cloud registration details are null, incomplete, or malformed
+     *
+     * @return cloud authentication results including the JWT token, owner ID, and or the anonymous
+     * consumer UUID
+     */
+    public CloudAuthenticationResultDTO generateRegistrationTokenV2(Principal principal,
+        CloudRegistrationInfo cloudRegInfo)
+        throws CloudRegistrationAuthorizationException, MalformedCloudRegistrationException {
+
+        if (!this.enabled) {
+            throw new UnsupportedOperationException(
+                "cloud registration is not enabled on this Candlepin instance");
+        }
+
+        if (principal == null) {
+            throw new IllegalArgumentException("principal is null");
+        }
+
+        if (cloudRegInfo == null) {
+            throw new IllegalArgumentException("cloudRegistrationInfo is null");
+        }
+
+        CloudAuthenticationResult result = this.cloudRegistrationAdapter
+            .resolveCloudRegistrationDataV2(cloudRegInfo);
+
+        String cloudAccountId = result.getCloudAccountId();
+        if (cloudAccountId == null || cloudAccountId.isEmpty() || cloudAccountId.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("cloud provider account could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String cloudProviderShortName = result.getCloudProviderShortName();
+        if (cloudProviderShortName == null || cloudProviderShortName.isEmpty() ||
+            cloudProviderShortName.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("cloud provider short name could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String productId = result.getProductId();
+        if (productId == null || productId.isEmpty() || productId.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("product id could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String ownerKey = result.getOwnerKey();
+        boolean ownerExistsAndPoolExists = false;
+        if (ownerKey != null && !ownerKey.isEmpty() && result.isEntitled()) {
+            Owner owner = ownerCurator.getByKey(ownerKey);
+            if (owner != null) {
+                ownerExistsAndPoolExists = !poolCurator.listByOwnerAndProduct(owner, productId).isEmpty();
```

Comment:
It's not strictly necessary to do both lookups in this case. If the org doesn't exist, the pool lookup will fail even if the product does, making the owner fetching redundant.

Reply:
Nice!! I'd rather just make the one curator call. Updated